core/src/main/java/io/grpc/transport/AbstractStream.java

```diff
@@ -351,28 +345,27 @@ protected final void onSendingBytes(int numBytes) {
   protected final void onSentBytes(int numBytes) {
     boolean doNotify;
     synchronized (onReadyLock) {
+      doNotify = wentBelowThreshold(
+          numSentBytesQueued,  // Previous queued sent bytes
+          onReadyThreshold,
+          numSentBytesQueued - numBytes, // New queued sent bytes
+          onReadyThreshold);
       numSentBytesQueued -= numBytes;
-      doNotify = needToNotifyOnReady();
     }
     if (doNotify) {
-      listener().onReady();
+      maybeNotifyListener();
```

Comment:
hmm ... don't you still have the same potential race with `maybeNotifyListener` in that you release the lock before calling the listener?